Dockerfile

```diff
@@ -0,0 +1,69 @@
+ARG FROM_IMAGE=ros:dashing
+FROM $FROM_IMAGE
+
+# install CI dependencies	
+RUN apt-get update && apt-get install -q -y \	
+      ccache \	
+    && rm -rf /var/lib/apt/lists/*
+
+# clone underlay package repos
+ENV ROS_WS /opt/ros_ws
+RUN mkdir -p $ROS_WS/src
+WORKDIR $ROS_WS
+COPY ./docker/underlay.repos ./
+RUN vcs import src < underlay.repos
+
+# install underlay package dependencies
+RUN apt-get update && \
+    rosdep install -q -y \
+      --from-paths \
+        src \
+      --ignore-src \
+    && rm -rf /var/lib/apt/lists/*
+
+# build underlay package source
+ARG CMAKE_BUILD_TYPE=Release
+RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
+    colcon build \
+      --symlink-install \
+      --cmake-args \
+        -DCMAKE_BUILD_TYPE=$CMAKE_BUILD_TYPE \
+        -DCMAKE_C_COMPILER_LAUNCHER=ccache \
+        -DCMAKE_CXX_COMPILER_LAUNCHER=ccache
```

Comment:
We could use some mixins here (same below):
https://github.com/colcon/colcon-mixin-repository/blob/master/ccache.mixin
https://github.com/colcon/colcon-mixin-repository/blob/master/build-type.mixin

Reply:
Do mixins 'mix well' with environment variables and build args?